### ISSUE TITLE
@as and @select on p:option

### ISSUE THREAD

**Issue by xml-project** [xproc/3.0-specification#486]
Is it a static error if the evaluation of `@select` on p:option return a value that does not match the sequence-type in `@as`?

Option 1: Yes, we need to introduce a new static error for this.
Option 2: No: We wait until the default value of the option is used and raise XD0036 then.

**Comment by gimsieke** [comment]
To check the sequence type during static analysis would certainly be more anal, as implied by “analysis.” 
Jokes aside. Since only the `@select` attributes of static options and variables need to be evaluated during static analysis, I’d say we cannot require a processor to evaluate _all_ options statically in order to determine the return type for the default values. For static variables and options, I think a sequence type error should be raised during static analysis though.

**Comment by xml-project** [comment]
>  I’d say we cannot require a processor to evaluate all options statically in order to determine the return type for the default values.

Contrary to your reading, I would say the processor has not just to determine the static type of `@select` but it's value:

> The select expression is evaluated during static analysis.

So the processor knows for all options (static or not) whether the default value matches the sequence type. The only question IMHO is whether this knowledge is used for a static error or not.

What did I miss?

**Comment by gimsieke** [comment]
In http://spec.xproc.org/master/head/xproc/#p.option it says indeed “The select expression is evaluated during static analysis.” It seems to apply to all options. 

On the other hand, there is http://spec.xproc.org/master/head/xproc/#statics “The values of static options and variables are computed during static analysis.” and http://spec.xproc.org/master/head/xproc/#static-expressions “Several kinds of expressions are evaluated during static analysis: … The select expressions on static p:options, unless a static default value has been provided.”
I think if all select expressions need to be evaluated anyway during static analysis, it need not be mentioned that the select expressions of static options need to be evaluated during static analysis.

Another indication the the rule “evaluate select during static analysis“ is not intended to be this general: The first sentence of the para that immediately precedes “The select expression is evaluated during static analysis.” reads “A p:option with a may be declared “static”.” Maybe “with a” is superfluous here? I can imagine that if the editing has been sloppy in the preceding para, then the sentence about evaluation of all options’ select expressions during static analysis is not intended to be this general, but it should only apply to static options. This assumption is corroborated by the fact that it was added in the same [commit](https://github.com/xproc/3.0-specification/blame/master/xproc/src/main/xml/specification.xml#L4899) as the static option stuff. Norm will be able to clarify this.

**Comment by xml-project** [comment]
As I said in the telcon, the specs are currently in a very complicated state to implement because different passage contradict each other and some phrases are hard to understand. E.g. I do not understand the meaning of the phrase you quoted:

> Several kinds of expressions are evaluated during static analysis:
(1) The select expressions on static p:options, unless a static default value has been provided.

I am not sure, how you read (1). To me the subclause can not refer to our problem of giving default value for a dynamic option because the main sentence is about "static option". Apart from this, I have no idea what this sentence says. 

I thought we agreed on default option values to be declared static, so 

> “The select expression is evaluated during static analysis.”

represents our consensus, that the default value of dynamic options is determined statically. But as you question whether is sentence really applies to `@select` on dynamic options, it obviously a mistake. 

We have to see, what @ndw has to say about it. In the meantime could you please object against my PR #485 because I need to change the error text if your reading is right.

**Comment by xml-project** [comment]
Two more pieces for the puzzle:

First:
> It is a static error (err:XS0092) if a p:with-option refers to an option that is declared static.

One could take this to mean, that option declared to be static are not visible during dynamic analysis. If this is right, `@select` on `p:option static="false"` either has to be evaluated statically (as the sentence in question says), or can not refer to an option declared as static.

Second:
In XProc 1.0 I can do this:
````
<p:option name="option1" required="true" />
<p:option name="option2" select="concat($option1,'-add')" />
````
This would be no longer valid according to the "general reading" of the sentence in question, because the select expression for option2 can not be evaluated statically. This would support @gimsieke 's interpretation, that 

> “The select expression is evaluated during static analysis.”

only refers to options declared as static.

**Comment by ndw** [comment]
RIght. I think I've muddied the waters significantly in trying to address static vs dynamic analysis. With respect to the assertion that "the select expression is evaluated during static analysis" in the description of `p:option`, I think that was just a mistake on my part. The 1.0 use case that @xml-project quotes above proves I was wrong. I'll try to clean things up.

**Comment by ndw** [comment]
Closed by #499
